net/host.go

```diff
@@ -79,25 +85,37 @@ func (h *Host) P2pHost() P2pHost {
 	return h.h
 }
 
-// SetHandler sets the Handler instance used by the host.
-func (h *Host) SetHandler(handler Handler) {
+// SetHandlers sets the maker and taker instances used by the host.
+func (h *Host) SetHandlers(makerHandler MakerHandler, takerHandler TakerHandler) {
+	h.makerHandler = makerHandler
+	h.takerHandler = takerHandler
+
 	fn := func() bool {
-		return len(handler.GetOffers()) == 0
+		return len(makerHandler.GetOffers()) > 0 || h.isRelayer
 	}
-
-	h.handler = handler
 	h.h.SetShouldAdvertiseFunc(fn)
 }
 
 // Start starts the bootstrap and discovery process.
 func (h *Host) Start() error {
-	if h.handler == nil {
+	if h.makerHandler == nil || h.takerHandler == nil {
 		return errNilHandler
 	}
 
-	h.h.SetStreamHandler(queryID, h.handleQueryStream)
+	h.h.SetStreamHandler(queryProtocolID, h.handleQueryStream)
+	h.h.SetStreamHandler(relayProtocolID, h.handleRelayStream)
 	h.h.SetStreamHandler(swapID, h.handleProtocolStream)
-	return h.h.Start()
+
+	// Note: Start() is non-blocking
+	if err := h.h.Start(); err != nil {
+		return err
+	}
+
+	if h.isRelayer {
+		h.Advertise(nil)
```

Comment:
If we don't call it here, we won't advertise that we are a relayer until the first XMR offer is made. Proposed fix for this was mentioned earlier.